src/Microsoft.SqlTools.ServiceLayer/QueryExecution/Batch.cs

```diff
@@ -84,11 +83,22 @@ internal Batch(string batchText, SelectionData selection, int ordinalId, IFileSt
         public delegate Task BatchAsyncEventHandler(Batch batch);
 
         /// <summary>
+        /// Asynchronous handler for when a message is emitted by the sql connection
+        /// </summary>
+        /// <param name="message">The message that was emitted</param>
+        public delegate Task BatchAsyncMessageHandler(ResultMessage message);
+
+        /// <summary>
         /// Event that will be called when the batch has completed execution
         /// </summary>
         public event BatchAsyncEventHandler BatchCompletion;
 
         /// <summary>
+        /// Event that will be called when a message has been emitted
+        /// </summary>
+        public event BatchAsyncMessageHandler BatchMessage;
```

Comment:
I think the better name would be BatchMessageSent, BatchMessage doesn't really say why happened to the message.